src/ctl/ctl.c

```diff
@@ -0,0 +1,593 @@
+    default:
+        break;
+    }
+
+    return real_arg;
+}
+
+/*
+ * ctl_query_cleanup_real_args -- (internal) cleanups relevant argument
+ *	structures allocated as a result of the get_real_args call
+ */
+static void ctl_query_cleanup_real_args(const struct ctl_node *n,
+                                        void *real_arg,
+                                        enum ctl_query_source source) {
+    /* suppress unused-parameter errors */
+    (void)n;
+
+    switch (source) {
+    case CTL_QUERY_CONFIG_INPUT:
+        Free(real_arg);
+        break;
+    case CTL_QUERY_PROGRAMMATIC:
+        break;
+    default:
+        break;
+    }
+}
+
+/*
+ * ctl_exec_query_read -- (internal) calls the read callback of a node
+ */
+static int ctl_exec_query_read(void *ctx, const struct ctl_node *n,
+                               enum ctl_query_source source, void *arg,
+                               struct ctl_index_utlist *indexes) {
+    if (arg == NULL) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    return n->cb[CTL_QUERY_READ](ctx, source, arg, indexes);
+}
+
+/*
+ * ctl_exec_query_write -- (internal) calls the write callback of a node
+ */
+static int ctl_exec_query_write(void *ctx, const struct ctl_node *n,
+                                enum ctl_query_source source, void *arg,
+                                struct ctl_index_utlist *indexes) {
+    if (arg == NULL) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    void *real_arg = ctl_query_get_real_args(n, arg, source);
+    if (real_arg == NULL) {
+        return -1;
+    }
+
+    int ret = n->cb[CTL_QUERY_WRITE](ctx, source, real_arg, indexes);
+    ctl_query_cleanup_real_args(n, real_arg, source);
+
+    return ret;
+}
+
+/*
+ * ctl_exec_query_runnable -- (internal) calls the run callback of a node
+ */
+static int ctl_exec_query_runnable(void *ctx, const struct ctl_node *n,
+                                   enum ctl_query_source source, void *arg,
+                                   struct ctl_index_utlist *indexes) {
+    return n->cb[CTL_QUERY_RUNNABLE](ctx, source, arg, indexes);
+}
+
+static int (*ctl_exec_query[MAX_CTL_QUERY_TYPE])(
+    void *ctx, const struct ctl_node *n, enum ctl_query_source source,
+    void *arg, struct ctl_index_utlist *indexes) = {
+    ctl_exec_query_read,
+    ctl_exec_query_write,
+    ctl_exec_query_runnable,
+};
+
+/*
+ * ctl_query -- (internal) parses the name and calls the appropriate methods
+ *	from the ctl tree
+ */
+int ctl_query(struct ctl *ctl, void *ctx, enum ctl_query_source source,
+              const char *name, enum ctl_query_type type, void *arg) {
+    if (name == NULL) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    /*
+	 * All of the indexes are put on this list so that the handlers can
+	 * easily retrieve the index values. The list is cleared once the ctl
+	 * query has been handled.
+	 */
+    struct ctl_index_utlist *indexes = NULL;
+    indexes = Zalloc(sizeof(*indexes));
```

Comment:
Check for `NULL`

Reply:
+1 @KFilipek please note that this kind of potential issue would be cached by Coverity

Reply:
Done